File: backend/app/core/features.py
FEATURE_CATALOG: list[dict] = [
    {"key": "mpesa_manual",        "label": "M-Pesa Manual Entry",      "group": "Payments",   "description": "Accept M-Pesa with manual reference entry"},
    {"key": "mpesa_stk",           "label": "M-Pesa STK Push",          "group": "Payments",   "description": "Automatically trigger M-Pesa payment prompts"},
    {"key": "sms_receipts",        "label": "SMS Receipts",             "group": "Receipts",   "description": "Send receipts via SMS to customers"},
    {"key": "credit_system",       "label": "Credit System",            "group": "Sales",      "description": "Issue invoices and track customer credit balances"},
    {"key": "advanced_reports",    "label": "Advanced Reports",         "group": "Analytics",  "description": "Full analytics: sales, inventory, products, credit"},
    {"key": "inventory_analytics", "label": "Inventory Analytics",      "group": "Analytics",  "description": "Reorder suggestions and aging stock reports"},
    {"key": "audit_logs",          "label": "Audit Logs",               "group": "Security",   "description": "Track all user actions for compliance"},
    {"key": "permissions_mgmt",    "label": "Custom Permissions",       "group": "Security",   "description": "Configure role-based access control per role"},
    {"key": "expenditure_tracking", "label": "Expenditure Tracking",     "group": "Finance",    "description": "Record and report on business expenses"},
    {"key": "multi_branch",        "label": "Multi-Branch",             "group": "Operations", "description": "Manage multiple store locations"},
    {"key": "supplier_management", "label": "Supplier Management",      "group": "Operations", "description": "Manage suppliers and create purchase orders"},
    {"key": "barcode_mode",        "label": "Barcode Scanner",          "group": "Operations", "description": "Scan barcodes to add products at POS"},
    {"key": "custom_units",        "label": "Custom Product Units",     "group": "Operations", "description": "Define custom units of measure for products"},
    {"key": "thermal_printing",    "label": "Thermal Printing",         "group": "Operations", "description": "Direct USB/Serial ESC/POS thermal printer support"},
    {"key": "product_images",      "label": "Product Images",           "group": "Operations", "description": "Upload and display product photos in inventory"},
    {"key": "api_access",          "label": "API Access",               "group": "Developer",  "description": "REST API access for third-party integrations"},
    {"key": "attendance_tracking", "label": "Attendance Tracking",      "group": "HR",         "description": "Staff clock in/out and attendance reporting"},
]

ALL_FEATURE_KEYS: set[str] = {f["key"] for f in FEATURE_CATALOG}

# Most features are plan-gated; a few are on by default for all orgs
_ALWAYS_ON = {"thermal_printing", "barcode_mode"}

DEFAULT_FLAGS: dict[str, bool] = {
    f["key"]: (f["key"] in _ALWAYS_ON) for f in FEATURE_CATALOG
}


def resolve_flags(plan_features: dict | None) -> dict[str, bool]:
    """Merge plan features with defaults, ensuring all known keys are present."""
    base = dict(DEFAULT_FLAGS)
    if plan_features:
        for k, v in plan_features.items():
            if k in ALL_FEATURE_KEYS and isinstance(v, bool):
                base[k] = v
    return base

File: mobile/lib/core/features.dart
/// Mirror of the backend feature catalog (app/core/features.py). Used to render
/// the subscription feature list and gate functionality in the app.
class FeatureDef {
  final String key;
  final String label;
  final String group;
  const FeatureDef(this.key, this.label, this.group);
}

const kFeatureCatalog = <FeatureDef>[
  FeatureDef('mpesa_manual', 'M-Pesa Manual Entry', 'Payments'),
  FeatureDef('mpesa_stk', 'M-Pesa STK Push', 'Payments'),
  FeatureDef('sms_receipts', 'SMS Receipts', 'Receipts'),
  FeatureDef('credit_system', 'Credit System', 'Sales'),
  FeatureDef('advanced_reports', 'Advanced Reports', 'Analytics'),
  FeatureDef('inventory_analytics', 'Inventory Analytics', 'Analytics'),
  FeatureDef('audit_logs', 'Audit Logs', 'Security'),
  FeatureDef('permissions_mgmt', 'Custom Permissions', 'Security'),
  FeatureDef('expenditure_tracking', 'Expenditure Tracking', 'Finance'),
  FeatureDef('multi_branch', 'Multi-Branch', 'Operations'),
  FeatureDef('supplier_management', 'Supplier Management', 'Operations'),
  FeatureDef('barcode_mode', 'Barcode Scanner', 'Operations'),
  FeatureDef('custom_units', 'Custom Product Units', 'Operations'),
  FeatureDef('thermal_printing', 'Thermal Printing', 'Operations'),
  FeatureDef('product_images', 'Product Images', 'Operations'),
  FeatureDef('api_access', 'API Access', 'Developer'),
  FeatureDef('attendance_tracking', 'Attendance Tracking', 'HR'),
];

// Feature keys used for gating in the app.
class Feat {
  static const mpesaManual = 'mpesa_manual';
  static const mpesaStk = 'mpesa_stk';
  static const creditSystem = 'credit_system';
  static const advancedReports = 'advanced_reports';
  static const multiBranch = 'multi_branch';
  static const barcodeMode = 'barcode_mode';
}
